File: app/src/main/java/com/example/android/popularmovies/activity/MainActivity.java
package com.example.android.popularmovies.activity;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.example.android.popularmovies.R;
import com.example.android.popularmovies.adapter.MoviesAdapter;
import com.example.android.popularmovies.model.Movie;
import com.example.android.popularmovies.model.MoviesResponse;
import com.example.android.popularmovies.rest.ApiClient;
import com.example.android.popularmovies.rest.ApiInterface;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private static String API_KEY;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        API_KEY = getString(R.string.api_key);

        final RecyclerView recyclerView = (RecyclerView) findViewById(R.id.movies_recycler_view);

        GridLayoutManager gridLayoutManager = new GridLayoutManager(getApplicationContext(), 3);
        recyclerView.setLayoutManager(gridLayoutManager);


        ApiInterface service = ApiClient.getClient().create(ApiInterface.class);

        Call<MoviesResponse> call = service.getPopularMovies(API_KEY);
        call.enqueue(new Callback<MoviesResponse>() {
            @Override
            public void onResponse(Call<MoviesResponse> call, Response<MoviesResponse> response) {

                List<Movie> movies = null;
                if (response.body() != null) {
                    movies = response.body().getResults();
                    recyclerView.setAdapter(new MoviesAdapter(movies, R.layout.list_items, getApplicationContext()));
                } else {
                    showError();
                }
            }

            @Override
            public void onFailure(Call<MoviesResponse> call, Throwable t) {

            }
        });
    }

    private void showError() {
        Toast.makeText(this, "Add your API address", Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/kotlin/com/example/android/popularmovies/activity/MainActivity.kt
package com.example.android.popularmovies.activity

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Context
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.widget.Toast
import com.example.android.popularmovies.R
import com.example.android.popularmovies.adapter.MoviesAdapter
import com.example.android.popularmovies.databinding.ActivityMainBinding
import com.example.android.popularmovies.model.Movie
import com.example.android.popularmovies.view_model.MainViewModel
import com.example.android.popularmovies.view_model.MainViewModelFactory
import java.util.*


class MainActivity : AppCompatActivity() {
    private var mContext: Context? = null

    private var mMovies = ArrayList<Movie>()
    private lateinit var moviesAdapter: MoviesAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        API_KEY = getString(R.string.api_key)
        mContext = this
        moviesAdapter = MoviesAdapter(mMovies)

        val activityMainBinding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        activityMainBinding.adapter = moviesAdapter

        val mainViewModel = ViewModelProviders.of(this, MainViewModelFactory(API_KEY!!)).get(MainViewModel::class.java)

        mainViewModel.getMovies().observe(this, Observer { movies ->
            if (movies != null) {
                moviesAdapter.addItem(movies)
            } else {
                Toast.makeText(mContext, R.string.error, Toast.LENGTH_LONG).show()
            }
        })
    }

    companion object {

        private var API_KEY: String? = null
    }
}
